### ISSUE TITLE
[Performance]  Getting different result in C# ONNX runtime and Python ONNX runtime

### ISSUE THREAD

**Issue by Jaswanth987** [microsoft/onnxruntime#19226]
### Describe the issue

I am using Python ONNX Runtime and loading **YOLOv8 ONNX model with NMS**(Non Max Suppression) inside it ,i am getting correct results in python , but when i use  **C#** ONNX Runtime 1.16.3 With GPU and  **.NET** Framework 4.6.1 i am not getting correct results.
May i know the possible reasons and solutions to it.
Thanks in Advance!!!

### Python Code :-

```python
onnx_model_path ="<path_to_onnx_model>
session = ort.InferenceSession(onnx_model_path)
# Load an image for inference
image_path = " <path_to_image>"
single_channel_image = cv2.imread(image_path,-1)
im_height, im_width,ch = input_image.shape
three_channel_image = cv2.cvtColor(single_channel_image, cv2.COLOR_GRAY2BGR)
input_image = three_channel_image / 255.0
input_data = np.expand_dims(input_data.astype(np.float32), axis=0)  # Add batch dimension
input_data = input_data.transpose(0,3,1,2)
outputs = ["det_bboxes", "det_scores", "det_classes", "num_dets"]
result = session.run(outputs, {"images": input_data})
```

### In C#:-

```C#
var session = new InferenceSession(object_detection_model,SessionOptions.MakeSessionOptionWithCudaProvider(gpuDeviceId));
Session = session;
inputName = session.InputMetadata.Keys.First();
outputName = session.OutputMetadata.Keys.First();
Mat inputImage = CvInvoke.Imread(resizeGrayImgPath, ImreadModes.Grayscale);
CvInvoke.CvtColor(inputImage, inputImage, ColorConversion.Gray2Bgr);
inputImage.ConvertTo(inputImage, DepthType.Cv32F, 1.0 / 255.0);
float[] inputArray = new float[inputImage.NumberOfChannels * inputImage.Width * inputImage.Height];
int count = 0;
for (int y = 0; y < inputImage.Height; y++)
{
    for (int x = 0; x < inputImage.Width; x++)
    {
        float[] pixelValues = new float[inputImage.NumberOfChannels];
        // Assuming the image is of type Mat
        using (var subRect = new Mat(inputImage, new Rectangle(x, y, 1, 1)))
        {
            float[] pixelData = new float[inputImage.NumberOfChannels];
            subRect.CopyTo(pixelData);
            for (int c = 0; c < inputImage.NumberOfChannels; c++)
            {
                float pixelValue = pixelData[c];
                inputArray[count++] = pixelValue;
            }
        }
    }
}
var inputTensor = new DenseTensor<float>(inputArray, new[] { 1, 3, inputImage.Height, inputImage.Width });
var inputs = new NamedOnnxValue[] { NamedOnnxValue.CreateFromTensor(inputName, inputTensor) };
var results = Session.Run(inputs).ToArray();
```

The initial input is uint8 1 channel gray scale image of size width 1024 and height 512  and we are converting this uint8 1 channel to float32 3 channel and sending these to Session.run to get outputs.

### Urgency

Yes, this issue is urgent, please help ASAP.

### Platform

Windows

### OS Version

Windows 11

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.16.3

### ONNX Runtime API

C#

### Architecture

X64

### Execution Provider

CUDA

### Is this a quantized model?

No

**Comment by tianleiwu** [comment]
You can build onnxruntime from source by adding `--cmake_extra_defines onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS=1`. Then you can set environment variables `ORT_DEBUG_NODE_IO_DUMP_INPUT_DATA` = 1 and  `ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA` = 1. When you run your application, it will dump the inputs and outputs of each node to Console. In this way, you can check whether the inputs are same in C# and python.

I guess some inputs are different for some reason.

**Comment by Jaswanth987** [comment]
Sorry, I forgot to mention above that inputs are same in C# and python , even after converting into float32 3 channel the values are same in both python and C# , but results are different in C# ONNX Runtime and Python ONNX Runtime.
_What might be possible reasons in this case._
Thanks in advance!!!

**Comment by tianleiwu** [comment]
If you redirect the dumped inputs/outputs to a text file, you can use VS code to compare the results of C# and Python to find out which inputs/node caused the difference. That's the standard way for such analysis.

Assume that you have the source model (like PyTorch to Tensorflow), which output (C# or Python) is same as source model for same input?

**Comment by Jaswanth987** [comment]
I have tried using single channel trained model in C# ONNX Runtime ,instead of three channel trained model , i am getting results same as python in case of single channel model, but using three channel model we are not getting same results as Python ONNX Runtime. 
What might be possible reasons and solutions to it?
Thanks in Advance!!!

**Comment by Craigacp** [comment]
Isn't the C# code emitting channels last and the Python code using channels first?

**Comment by xadupre** [comment]
Just to make sure, your python code include ``input_data = np.expand_dims(input_data.astype(np.float32), axis=0)  # Add batch dimension`` but I would expect ``input_data = np.expand_dims(input_image.astype(np.float32), axis=0)  # Add batch dimension``. Another difference is you seem to run the python code on cpu (``ort.InferenceSession(onnx_model_path)``) and the C# code on GPU. One way to detect where discrepancies come from is to remove the one part of the model and compare intermediate result. Then you would know which node introduce the first errors.

**Comment by Jaswanth987** [comment]
In the above mentioned C# code , after getting float1D array then we are passing this to dense tensor of shape of **[1,3,inputImage.Height,InputImage.Width]** , instead of doing this we have taken float 1D array and  reshaped it to  4D array of shape **[1,inputImage.Height,InputImage.Width,3]** (which is shape of Mat inputImage **[512,1024,3]**  and adding batch dimension to it)and then we transposed this 4D array to **[1,3,inputImage.Height,InputImage.Width]** and then converted into float 1D array and then passed this float 1D array to DenseTensor and passed this DenseTensor to Sessions.Run and then **we got outputs same as python**.

**Comment by Jaswanth987** [comment]
In above C# code does this line 
`var inputTensor = new DenseTensor<float>(inputArray, new[] { 1, 3, inputImage.Height, inputImage.Width });`
automatically does reshape and then transpose?

**Comment by xwhkkk** [comment]
> In the above mentioned C# code , after getting float1D array then we are passing this to dense tensor of shape of **[1,3,inputImage.Height,InputImage.Width]** , instead of doing this we have taken float 1D array and reshaped it to 4D array of shape **[1,inputImage.Height,InputImage.Width,3]** (which is shape of Mat inputImage **[512,1024,3]** and adding batch dimension to it)and then we transposed this 4D array to **[1,3,inputImage.Height,InputImage.Width]** and then converted into float 1D array and then passed this float 1D array to DenseTensor and passed this DenseTensor to Sessions.Run and then **we got outputs same as python**.

Thank you for your suggestion. I would like to ask how to pass the float 1D DenseTensor to Sessions.Run. The following code `var inputTensor = new DenseTensor<float>(inputArray, new[] { 1, 3, inputImage.Height, inputImage.Width });` 
would automatically reshape the tensor in 4D. The input value of onnxSession usually has fixed input dimensions.
